File: scripts/coinalyze_api.py
from typing import List, Dict, Any, Optional
import time
import requests
from dataclasses import dataclass
from urllib.parse import urljoin

@dataclass
class HistoricalParams:
    symbols: List[str]
    interval: str
    from_timestamp: int
    to_timestamp: int
    convert_to_usd: bool = False


 
class CoinAlyzeAPI:
    def __init__(
        self,
        api_key: str,
        base_url: str = "https://api.coinalyze.net/v1/",
        rate_limit: int = 40
    ):
        self.api_key = api_key
        self.base_url = base_url
        self.rate_limit = rate_limit
        self.last_request_time = 0

    def _make_request(self, endpoint: str, params: Optional[Dict[str, Any]] = None) -> Any:
        """
        Base request handler with rate limiting
        """
        now = time.time() * 1000  # Convert to milliseconds
        time_since_last = now - self.last_request_time

        # Enforce rate limit (40 requests/minute)
        if time_since_last < 1500:  # 60s/40 = 1.5s between requests
            time.sleep((1500 - time_since_last) / 1000)  # Convert back to seconds

        url = urljoin(self.base_url, endpoint)

        if params:
            # Convert list parameters to comma-separated strings
            processed_params = {
                k: ','.join(v) if isinstance(v, list) else str(v)
                for k, v in params.items()
            }
        else:
            processed_params = {}

        headers = {
            'x-api-key': self.api_key,
            'Content-Type': 'application/json'
        }

        response = requests.get(url, headers=headers, params=processed_params)
        self.last_request_time = time.time() * 1000

        if not response.ok:
            if response.status_code == 429:
                retry_after = int(response.headers.get('Retry-After', '60'))
                time.sleep(retry_after)
                return self._make_request(endpoint, params)
            raise Exception(f"API Error {response.status_code}: {response.text}")

        return response.json()

    def get_supported_exchanges(self) -> List[Dict[str, str]]:
        """
        Get list of supported exchanges
        """
        return self._make_request('exchanges')

    def get_current_funding_rates(self, symbols: List[str]) -> List[Dict[str, Any]]:
        """
        Get current funding rates for specified symbols
        """
        return self._make_request('funding-rate', {'symbols': symbols})

    def get_ohlcv_history(self, params: HistoricalParams) -> List[Dict[str, Any]]:
        """
        Get OHLCV history for specified parameters
        """
        request_params = {
            'symbols': params.symbols,
            'interval': params.interval,
            'from': params.from_timestamp,
            'to': params.to_timestamp
        }
        return self._make_request('ohlcv-history', request_params)

    def get_open_interest_history(self, params: HistoricalParams) -> List[Dict[str, Any]]:
        """
        Get open interest history for specified parameters
        """
        request_params = {
            'symbols': params.symbols,
            'interval': params.interval,
            'from': params.from_timestamp,
            'to': params.to_timestamp,
            'convert_to_usd': str(params.convert_to_usd).lower()
        }
        return self._make_request('open-interest-history', request_params) 
print(CoinAlyzeAPI.get_current_funding_rates(['BTCUSDT_PERP']))

File: src/scripts/coinalyze-api.ts
import fetch from 'node-fetch';

interface CoinAlyzeConfig {
  apiKey: string;
  baseUrl?: string;
  rateLimit?: number;
}

interface HistoricalParams {
  symbols: string[];
  interval: string;
  from: number;
  to: number;
  convertToUSD?: boolean;
}

interface ExchangeInfo {
  name: string;
  code: string;
}

interface FundingRate {
  symbol: string;
  value: number;
  update: number;
}

interface OHLCVData {
  symbol: string;
  history: Array<{
    t: number;  // timestamp
    o: number;  // open
    h: number;  // high
    l: number;  // low
    c: number;  // close
    v: number;  // volume
  }>;
}

interface OpenInterestData {
  symbol: string;
  history: Array<{
    t: number;    // timestamp
    value: number;
  }>;
}

export class CoinAlyzeAPI {
  private readonly config: Required<CoinAlyzeConfig>;
  private lastRequestTime = 0;
  
  constructor(config: CoinAlyzeConfig) {
    this.config = {
      baseUrl: 'https://api.coinalyze.net',
      rateLimit: 40,
      ...config
    };
  }

  /**
   * Base request handler with rate limiting
   */
  private async makeRequest<T>(endpoint: string, params?: Record<string, any>): Promise<T> {
    const now = Date.now();
    const timeSinceLast = now - this.lastRequestTime;
    
    // Enforce rate limit (40 requests/minute)
    if (timeSinceLast < 1500) { // 60s/40 = 1.5s between requests
      await new Promise(resolve => setTimeout(resolve, 1500 - timeSinceLast));
    }

    // Add /v1/ to the endpoint path
    const url = new URL(`${this.config.baseUrl}/v1/${endpoint}`);
    
    console.log('Making request to:', url.toString()); // Debug log
    
    if (params) {
      Object.entries(params).forEach(([key, value]) => {
        if (Array.isArray(value)) {
          url.searchParams.append(key, value.join(','));
        } else {
          url.searchParams.append(key, value.toString());
        }
      });
    }

    // Try different header variations
    const headers = {
      'X-API-Key': this.config.apiKey,  // Capitalized version
      'x-api-key': this.config.apiKey,  // lowercase version
      'apikey': this.config.apiKey,     // simple version
      'Accept': 'application/json'
      // Removed Content-Type as it's a GET request
    };

    console.log('Request headers:', headers); // Debug log

    const response = await fetch(url.toString(), {
      method: 'GET',
      headers
    });

    this.lastRequestTime = Date.now();

    if (!response.ok) {
      const errorText = await response.text();
      console.log('Full response:', {
        status: response.status,
        headers: Object.fromEntries(response.headers.entries()),
        body: errorText,
        url: url.toString() // Log the URL that was called
      });
      throw new Error(`API Error ${response.status}: ${errorText}`);
    }

    return response.json() as Promise<T>;
  }

  /**
   * Get list of supported exchanges
   */
  async getSupportedExchanges(): Promise<ExchangeInfo[]> {
    return this.makeRequest<ExchangeInfo[]>('exchanges');
  }

  /**
   * Get current funding rates for specified symbols
   */
  async getCurrentFundingRates(symbols: string[]): Promise<FundingRate[]> {
    return this.makeRequest<FundingRate[]>('funding-rate', { symbols });
  }

  /**
   * Get OHLCV history for specified parameters
   */
  async getOHLCVHistory(params: HistoricalParams): Promise<OHLCVData[]> {
    return this.makeRequest<OHLCVData[]>('ohlcv-history', {
      symbols: params.symbols,
      interval: params.interval,
      from: params.from,
      to: params.to
    });
  }

  /**
   * Get open interest history for specified parameters
   */
  async getOpenInterestHistory(params: HistoricalParams): Promise<OpenInterestData[]> {
    return this.makeRequest<OpenInterestData[]>('open-interest-history', {
      symbols: params.symbols,
      interval: params.interval,
      from: params.from,
      to: params.to,
      convert_to_usd: params.convertToUSD ? 'true' : 'false'
    });
  }
}

// Example usage:
/*
async function example() {
  const api = new CoinAlyzeAPI({
    apiKey: 'your-api-key-here'
  });

  // Get supported exchanges
  const exchanges = await api.getSupportedExchanges();
  console.log('Supported exchanges:', exchanges);

  // Get current funding rates
  const fundingRates = await api.getCurrentFundingRates(['BTC-PERP', 'ETH-PERP']);
  console.log('Funding rates:', fundingRates);

  // Get OHLCV history
  const ohlcvData = await api.getOHLCVHistory({
    symbols: ['BTC-PERP'],
    interval: '1h',
    from: Date.now() - 24 * 60 * 60 * 1000, // 24 hours ago
    to: Date.now()
  });
  console.log('OHLCV data:', ohlcvData);
}
*/ 